src/main/java/com/google/api/codegen/transformer/OutputTransformer.java

```diff
@@ -211,51 +222,104 @@ private static int identifier(String s, int startFrom) {
+                    .getAndSaveTypeName(context.getTypeTable(), scope.get(loop.getVariable())))
             .variableName(context.getNamer().localVarName(Name.from(loop.getVariable())))
             .collection(accessor)
             .body(
                 loop.getBodyList()
                     .stream()
-                    .map(body -> toView(body, context, valueSet, localVars))
+                    .map(body -> toView(body, context, valueSet, scope))
                     .collect(ImmutableList.toImmutableList()))
             .build();
 
     // The variable is only visible within the loop, delete it from the table before we return.
-    localVars.remove(loop.getVariable());
     return ret;
   }
 
-  private static OutputView.AssignmentView assignmentView(
-      String assign,
-      MethodContext context,
-      SampleValueSet valueSet,
-      Map<String, TypeModel> localVars) {
-
-    int p = identifier(assign, 0);
-    String ident = assign.substring(0, p);
+  private static OutputView.DefinitionView definitionView(
+      String definition, MethodContext context, SampleValueSet valueSet, ScopeTable localVars) {
+    int p = identifierEnd(definition, 0);
+    String ident = definition.substring(0, p);
     Preconditions.checkArgument(
-        assign.startsWith("=", p),
-        "%s:%s invalid assign, expecting '=': %s",
+        definition.startsWith("=", p),
+        "%s:%s invalid definition, expecting '=': %s",
         context.getMethodModel().getSimpleName(),
         valueSet.getId(),
-        assign);
+        definition);
     OutputView.VariableView reference =
-        accessor(assign.substring(p + 1), context, valueSet, localVars, ident);
-    return OutputView.AssignmentView.newBuilder()
+        accessorNewVariable(definition.substring(p + 1), context, valueSet, localVars, ident);
+    return OutputView.DefinitionView.newBuilder()
         .variableType(
             context.getNamer().getAndSaveTypeName(context.getTypeTable(), localVars.get(ident)))
         .variableName(context.getNamer().localVarName(Name.from(ident)))
         .reference(reference)
         .build();
   }
+
+  /**
+   * We track two scopes: universe and local.
+   *
+   * <p>Universe keeps track of all variable declared by the output specs. We need this because
+   * variables are function-scoped in many dynamic languages, and we should error if the spec
+   * declares variable with the same name twice.
+   *
+   * <p>Local keeps track of variable in the current block. We need this because variables are
+   * block-scoped in many static languages, and we should error if the spec uses a variable not in
+   * the current block or its transitive parents.
+   */
+  private static class ScopeTable {
+    private final Set<String> universe;
+    @Nullable private final ScopeTable parent;
+    private final Map<String, TypeModel> scope = new HashMap<>();
+
+    private ScopeTable() {
```

Comment:
TIL that private methods in private inner classes get package-level access: https://stackoverflow.com/questions/663059/why-do-inner-classes-make-private-methods-accessible